### PR TITLE
Storage Standardization [Take 2]

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Standardizes items to a 99% more sane system.

Sizes are now an enum (Tiny, Small, Normal, Large, Huge, Ginormous) which determine what kinds of containers an item can fit in. For example, large items can't fit in small containers. Sizes also have a numerical value which can be seen in the value (though this is pending to change in the future) which mimics the existing system.

This also comes with an additional lever, called "slots." **SLOTS ARE ONLY USED BY BELTS IN THE CURRENT DESIGN.** Slots simply specify a max number of items that can fit and ignores cumulative item sizes.

Naturally, a lot of values have been tweaked so balance might be fucky. This can all be changed rather trivially so don't sweat it too hard, capiche?

## Why / Balance

The purely numerical storage system had two main issues:

**Thresholds**
A lot of the values were built around numerical size thresholds that corresponded to in-game containers. These include:
- size 12: pockets
- size 100: backpacks
- size 120: duffels
- size 9999 (yes this was in the game fml): BoH

By using these standardized enum values instead of numbers, these items can more directly reflect the actual thresholds they're balanced around while simplifying the amount of numbers being juggled around.

**Cumulative Sizes**
10 size 1 items vs. 1 size 10 item. It's just not the same. Containers were not able to handle storing many smaller items without additionally storing larger ones than intended, barring you go through the pain of adding whitelists to it. The simple fact is that this makes balancing small containers like pill bottles very garbage.

Max item sizes can block this unwanted behavior, for example, by restricting Large items from being placed in toolbelts or Small items from being placed in matchboxes. This will allow coders more ways to make multipurpose and dynamic storage options (epical!)

## Technical details

**ItemComponent**
Size - Enum representing a different rough estimate of size.

**StorageComponent**
MaxTotalWeight - replacement for capacity, integer. Required for all storage unless they use slots.
MaxSlots - optional: alternative system that dictates a max number of items that can be placed
MaxItemSize - optional: the max size item that can be inserted into this storage. If null, it will be generated using an appropriate default value.

## Media

image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

Same shit as #21212

**Changelog**

:cl:
- add: Storage containers can now block large items from being inserted into them, regardless of the space available.
- add: Belts now function on a slot based storage system rather than a volume based one.
- tweak: Standardized item sizes into regular categories.

### CODE REVIEW

**Comment by mirrorcult** [comment]
much more based

**Comment by EmoGarbage404** [comment]
> much more based

one may say i encountered a brief lapse in based-ed-ness

**Comment by LankLTE** [comment]
I think this is better, but I'm still not a huge fan of allowing nested containers to give you more storage. If I'm going over this correctly, boxes weigh 4 and can hold 12, so you're still encouraging people to fill their backpacks with boxes to hold more stuff. Making "normal" items weigh 6 / hold 6 would essentially fix this while also being closer to the current storage system (one normal item of 30 weighing 6x a tiny item of 5), instead of making normal items significantly easier to carry in large amounts while also encouraging nested storage again (maybe even further, since you have more slots to put other stuff in excluding your nested storage).

**Comment by QuietlyWhisper** [comment]
> I think this is better, but I'm still not a huge fan of allowing nested containers to give you more storage. If I'm going over this correctly, boxes weigh 4 and can hold 12, so you're still encouraging people to fill their backpacks with boxes to hold more stuff. Making "normal" items weigh 6 / hold 6 would essentially fix this while also being closer to the current storage system (one normal item of 30 weighing 6x a tiny item of 5), instead of making normal items significantly easier to carry in large amounts while also encouraging nested storage again (maybe even further, since you have more slots to put other stuff in excluding your nested storage).

I think the intention is to give sub-storage entities value and use, by allowing them to carry more items. However this is better done with whitelisted items like pill canisters, which can be sized to be tiny/small but hold a lot of specific contents, rather than the bullshittery we saw with boxes in the discord forum...

It only gets stupid when you can fit multiple storage expansions into your bag that have no content restrictions it seems-

**Comment by UbaserB** [comment]
> I think this is better, but I'm still not a huge fan of allowing nested containers to give you more storage. If I'm going over this correctly, boxes weigh 4 and can hold 12, so you're still encouraging people to fill their backpacks with boxes to hold more stuff. Making "normal" items weigh 6 / hold 6 would essentially fix this while also being closer to the current storage system (one normal item of 30 weighing 6x a tiny item of 5), instead of making normal items significantly easier to carry in large amounts while also encouraging nested storage again (maybe even further, since you have more slots to put other stuff in excluding your nested storage).

i dont think encouraging boxes is a particularly bad thing. I mean for one, why would carrying 8837438 pills NOT in canisters make sense?

**Comment by JoeHammad1844** [comment]
I like this change, less hidden values are also nice, but what does a backpack with a medkit inside look like?

**Comment by JoeHammad1844** [comment]
> I think this is better, but I'm still not a huge fan of allowing nested containers to give you more storage. If I'm going over this correctly, boxes weigh 4 and can hold 12

the tradeoff is that you cant store normal sized items

**Comment by BasedUser** [comment]
Not particularly a fan, since this also breaks all forks' prototypes. However, many of the considerations from the storage discussion thread have been taken in.

**Comment by BasedUser** [comment]
From my limited playtesting (and intent to bring ss13 balance when the dust settles), I'd love to see an "arbitrage whitelist" be added, allowing you to put otherwise large (but not volume-exhausting) items in select containers (i.e. mops IN backpacks, if that were to be an intended feature).

Otherwise, LGTM.

> 7 stun batons in a belt is a valid state with this system.

You can hold 4 on /tg/, and that's with the security belt nerf.

**Comment by CatMagic1** [comment]
much better 👍

but also c4 matchboxes????
no more c4 matchboxes :(

**Comment by JoeHammad1844** [comment]
what do stacks look like?

**Comment by JoeHammad1844** [comment]
what just got merged that made this conflict

File: Content.Shared/Item/SharedItemSystem.cs [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
     [Dependency] private   readonly SharedHandsSystem _handsSystem = default!;
-    [Dependency] private   readonly SharedCombatModeSystem _combatMode = default!;
     [Dependency] protected readonly SharedContainerSystem Container = default!;
 
+    public const int ItemSizeWeightTiny = 1;
+    public const int ItemSizeWeightSmall = 2;
+    public const int ItemSizeWeightNormal = 4;
+    public const int ItemSizeWeightLarge = 16;
+    public const int ItemSizeWeightHuge = 24;
+    public const int ItemSizeWeightGinormous = 48;
```
Comment (mirrorcult) [inline Content.Shared/Item/SharedItemSystem.cs:23]: should this not just be the values of the enum

**Review by mirrorcult — CHANGES_REQUESTED** [review]

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**